src/cargo/ops/cargo_new.rs

```diff
@@ -548,6 +608,95 @@ fn global_config(config: &Config) -> CargoResult<CargoNewConfig> {
     })
 }
 
+/// Recursively list directory contents under `dir`, only visiting files.
+///
+/// This will also filter out files & files types which we don't want to
+/// try generate templates for. Image files, for instance.
+///
+/// It also filters out certain files & file types, as we don't want t
+///
+/// We use this instead of std::fs::walk_dir as it is marked as unstable for now
+///
+/// This is a modified version of the example at:
+///    http://doc.rust-lang.org/std/fs/fn.read_dir.html
+fn walk_template_dir(dir: &Path, cb: &mut FnMut(DirEntry)) -> CargoResult<()> {
+    let attr = try!(fs::metadata(&dir));
+
+    let ignore_files = vec!(".gitignore");
+    let ignore_types = vec!("png", "jpg", "gif");
+
+    if attr.is_dir() {
+        for entry in try!(fs::read_dir(dir)) {
+            let entry = try!(entry);
+            let attr = try!(fs::metadata(&entry.path()));
+            if attr.is_dir() {
+                if !&entry.path().to_str().unwrap().contains(".git") {
+                    try!(walk_template_dir(&entry.path(), cb));
+                }
+            } else {
+                if let &Some(extension) = &entry.path().extension() {
+                    if ignore_types.contains(&extension.to_str().unwrap()) {
+                        continue
+                    }
+                }
+                if let &Some(file_name) = &entry.path().file_name() {
+                    if ignore_files.contains(&file_name.to_str().unwrap()) {
+                        continue
+                    }
+                }
+                cb(entry);
+            }
+        }
+    }
+    Ok(())
+}
+
+/// Create a generic template
+///
+/// This consists of a Cargo.toml, and a src directory.
+fn create_generic_template(path: &PathBuf) -> CargoResult<()> {
+    match fs::metadata(&path) {
+        Ok(_) => {}
+        Err(_) => { try!(fs::create_dir(&path)); }
+    }
+    match fs::metadata(&path.join("src")) {
+        Ok(_) => {}
+        Err(_) => { try!(fs::create_dir(&path.join("src"))); }
+    }
+    try!(file(&path.join("Cargo.toml"), b"\
+[package]
+name = \"{{name}}\"
+version = \"0.1.0\"
+authors = [\"{{authors}}\"]
+"));
+    Ok(())
+}
+
+/// Create a new "lib" project
+fn create_lib_template(path: &PathBuf) -> CargoResult<()> {
+    try!(create_generic_template(&path));
+    try!(file(&path.join("src/lib.rs"), b"\
+#[test]
+fn it_works() {
+}
+"));
+    Ok(())
+}
+
+/// Create a new "bin" project
+fn create_bin_template(path: &PathBuf) -> CargoResult<()> {
```

Comment:
It seems like these should perhaps be "built in" rather than literally created on the filesystem?